File: Java/src/AESKeys.java
public class AESKeys {
	private AESMatrix[] keys;

	public AESKeys(byte[] key) {
		expandKey(key);
	}

	public AESMatrix get(int i) {
		return new AESMatrix(keys[i]);
	}

	private static final byte[] RCON = {
		0x01, (byte)0x02, (byte)0x04, (byte)0x08, (byte)0x10, (byte)0x20, (byte)0x40, (byte)0x80, (byte)0x1b, (byte)0x36
	};

	private void expandKey(byte[] key) {
		keys = new AESMatrix[11];
		keys[0] = new AESMatrix(key);

		for(int i = 0; i < 10; ++i) {
			keys[i+1] = genKey(keys[i], i);
		}
	}

	private AESMatrix genKey(AESMatrix previousKey, int rconN) {
		byte[] previousWord = new byte[4];
		byte[] newKey = new byte[16];

		// Apply rotation, and substitution on previous word with rcon xor
		previousWord[0] = previousKey.get(13);
		previousWord[1] = previousKey.get(14);
		previousWord[2] = previousKey.get(15);
		previousWord[3] = previousKey.get(12);
		AESMatrix.subByteArr(previousWord);
		previousWord[0] ^= RCON[rconN];

		for(int i = 0; i < 4; i++) {
			if(i != 0) {
				System.arraycopy(newKey, 4*(i-1), previousWord, 0, 4);
			}

			for(int j = 0; j < 4; ++j) {
				newKey[4 * i + j] = (byte)(previousWord[j] ^ previousKey.get(4 * i + j));
			}
		}

		return new AESMatrix(newKey);
	}
}

File: Lua/src/aes_keys.lua
local AESMatrix = require("src/aes_matrix")

AESKeys = {
	keys = {},
	RCON = {
		0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36
	};
}

function AESKeys:new()
	local new = {keys = {}}
	setmetatable(new, self)
	self.__index = self
	return new
end

function AESKeys:__tostring()
	local out = "\n======= START KEYS =======\n"
	for _, v in ipairs(self.keys) do
		out = out .. v:__tostring()
	end
	out = out .. "======== END KEYS ========\n"
	return out
end

function AESKeys:from_bytes(bytes)
	self.keys[1] = AESMatrix:new()
	self.keys[1]:from_bytes(bytes)

	for i, v in ipairs(self.RCON) do
		self:gen_key(i, v)
	end
end

function AESKeys:gen_key(key_i, rcon_v)
	local prev_key = self:get(key_i)
	local prev_word = {}
	local new_key = AESMatrix:new()

	-- copy rotated
	prev_word[1] = prev_key:get(14)
	prev_word[2] = prev_key:get(15)
	prev_word[3] = prev_key:get(16)
	prev_word[4] = prev_key:get(13)

	AESMatrix.sub_byte_arr(prev_word)
	prev_word[1] = prev_word[1] ~ rcon_v

	for i = 0, 3 do
		if i ~= 0 then
			for j = 0, 3 do
				prev_word[j+1] = new_key:get((i-1)*4+j+1)
			end
		end

		for j = 0, 3 do
			new_key:set(i*4+j+1, prev_word[j+1] ~ prev_key:get(i*4+j+1))
		end
	end

	self.keys[key_i + 1] = new_key
end

function AESKeys:get(n)
	return self.keys[n]
end


-- return statement
return AESKeys
